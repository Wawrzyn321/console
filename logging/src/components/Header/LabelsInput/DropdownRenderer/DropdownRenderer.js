import React from 'react';
import PropTypes from 'prop-types';
import './DropdownRenderer.scss';

DropdownRenderer.propTypes = {
  recentLabels: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  logLabels: PropTypes.arrayOf(PropTypes.object.isRequired).isRequired,
  chooseLabel: PropTypes.func.isRequired,
};

export default function DropdownRenderer({
  recentLabels,
  logLabels,
  chooseLabel,
}) {
  const [statefulLogLabels, setStatefulLogLabels] = React.useState(
    logLabels
      .filter(logLabel => logLabel.labels.length)
      .map(logLabel => {
        return {
          isHidden: true,
          logLabel: logLabel,
        };
      }),
  );

  function formatName(name) {
    return name.replace(' ', '_').toLowerCase();
  }

  function switchState(labelName) {
    setStatefulLogLabels(
      statefulLogLabels.map(entry => {
        if (entry.logLabel.name === labelName) {
          return {
            ...entry,
            isHidden: !entry.isHidden,
          };
        } else {
          return {
            ...entry,
            isHidden: true,
          };
        }
      }),
    );
  }

  const recentLabelsList = recentLabels.length ? (
    <ul className="fd-mega-menu__list">
      {recentLabels.map(entry => (
        <li key={entry} className="fd-mega-menu__item">
          <span
            className="cursor-pointer fd-mega-menu__link"
            onClick={() => chooseLabel(entry)}
          >
            {entry}
          </span>
        </li>
      ))}
    </ul>
  ) : (
    <span
      className="fd-mega-menu__item--disabled"
      data-test-id="no-recent-labels"
    >
      No recent labels
    </span>
  );

  const logLabelsSubList = ({ logLabel, isHidden }) => (
    <ul
      className="fd-mega-menu__sublist"
      id={logLabel.name}
      aria-hidden={isHidden}
    >
      {logLabel.labels.map(name => (
        <li className="fd-mega-menu__subitem" key={name}>
          <span
            className="fd-mega-menu__sublink cursor-pointer"
            onClick={() =>
              chooseLabel(`${formatName(logLabel.name)}="${name}"`)
            }
          >
            {name}
          </span>
        </li>
      ))}
    </ul>
  );

  const logLabelsList = logLabels.length ? (
    <ul className="fd-mega-menu__list">
      {statefulLogLabels.map(l => (
        <li className="fd-mega-menu__item" key={l.logLabel.name}>
          <span
            className="fd-mega-menu__link has-child cursor-pointer"
            aria-controls={l.logLabel.name}
            aria-haspopup="true"
            onClick={() => switchState(l.logLabel.name)}
          >
            {l.logLabel.name}
          </span>
          {logLabelsSubList(l)}
        </li>
      ))}
    </ul>
  ) : (
    <span className="fd-mega-menu__item--disabled" data-test-id="no-log-labels">
      No log labels
    </span>
  );

  return (
    <nav className="fd-mega-menu">
      <div className="fd-mega-menu__group">
        <h3 className="fd-mega-menu__title">Recently Selected Labels</h3>
        {recentLabelsList}
      </div>
      <div className="fd-mega-menu__group">
        <h3 className="fd-mega-menu__title">Log Labels</h3>
        {logLabelsList}
      </div>
    </nav>
  );
}
