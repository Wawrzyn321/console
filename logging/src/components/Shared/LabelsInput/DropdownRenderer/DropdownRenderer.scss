.fd-mega-menu__item--disabled {
    padding: 10px 20px;
    color: #51555a;
    display: block;
    font-size: 1rem;
    line-height: 1.42857;
    font-weight: 400;
    color: var(--fd-color-text-4);
}
