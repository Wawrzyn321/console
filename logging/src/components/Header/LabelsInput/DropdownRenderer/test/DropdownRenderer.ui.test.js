import React from 'react';
import { shallow } from 'enzyme';
import DropdownRenderer from './../DropdownRenderer';
import toJson from 'enzyme-to-json';

const mockRecentLabels = ['a', 'b'];

const mockLogLabels = [
  {
    name: 'name',
    labels: [['c', 'd']],
  },
  {
    name: 'name 2',
    labels: [['e', 'f']],
  },
];

describe('DropdownRenderer UI', () => {
  it('Shows / hides nested list when parent its parent is clicked', () => {
    const component = shallow(
      <DropdownRenderer
        recentLabels={[]}
        logLabels={mockLogLabels}
        chooseLabel={() => {}}
      />,
    );

    const link = component.find('span[aria-haspopup="true"]').first();

    link.simulate('click');
    expect(toJson(component)).toMatchSnapshot();

    link.simulate('click');
    expect(toJson(component)).toMatchSnapshot();
  });

  it('Click on subitem triggers callback', async () => {
    const mockCallback = jest.fn();
    const component = shallow(
      <DropdownRenderer
        recentLabels={mockRecentLabels}
        logLabels={[]}
        chooseLabel={mockCallback}
      />,
    );

    const link = component.find('span.fd-mega-menu__link').first();

    link.simulate('click');

    expect(mockCallback).toHaveBeenCalledTimes(1);
  });
});
