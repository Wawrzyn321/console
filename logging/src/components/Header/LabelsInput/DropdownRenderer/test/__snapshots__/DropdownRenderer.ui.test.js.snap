// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownRenderer Shows / hides nested list when parent its parent is clicked 1`] = `
<nav
  className="fd-mega-menu"
>
  <div
    className="fd-mega-menu__group"
  >
    <h3
      className="fd-mega-menu__title"
    >
      Recently Selected Labels
    </h3>
    <span
      className="fd-mega-menu__item--disabled"
      data-test-id="no-recent-labels"
    >
      No recent labels
    </span>
  </div>
  <div
    className="fd-mega-menu__group"
  >
    <h3
      className="fd-mega-menu__title"
    >
      Log Labels
    </h3>
    <ul
      className="fd-mega-menu__list"
    >
      <li
        className="fd-mega-menu__item"
        key="name"
      >
        <span
          aria-controls="name"
          aria-haspopup="true"
          className="fd-mega-menu__link has-child cursor-pointer"
          onClick={[Function]}
        >
          name
        </span>
        <ul
          aria-hidden={false}
          className="fd-mega-menu__sublist"
          id="name"
        >
          <li
            className="fd-mega-menu__subitem"
            key="c,d"
          >
            <span
              className="fd-mega-menu__sublink cursor-pointer"
              onClick={[Function]}
            >
              c
              d
            </span>
          </li>
        </ul>
      </li>
      <li
        className="fd-mega-menu__item"
        key="name 2"
      >
        <span
          aria-controls="name 2"
          aria-haspopup="true"
          className="fd-mega-menu__link has-child cursor-pointer"
          onClick={[Function]}
        >
          name 2
        </span>
        <ul
          aria-hidden={true}
          className="fd-mega-menu__sublist"
          id="name 2"
        >
          <li
            className="fd-mega-menu__subitem"
            key="e,f"
          >
            <span
              className="fd-mega-menu__sublink cursor-pointer"
              onClick={[Function]}
            >
              e
              f
            </span>
          </li>
        </ul>
      </li>
    </ul>
  </div>
</nav>
`;

exports[`DropdownRenderer Shows / hides nested list when parent its parent is clicked 2`] = `
<nav
  className="fd-mega-menu"
>
  <div
    className="fd-mega-menu__group"
  >
    <h3
      className="fd-mega-menu__title"
    >
      Recently Selected Labels
    </h3>
    <span
      className="fd-mega-menu__item--disabled"
      data-test-id="no-recent-labels"
    >
      No recent labels
    </span>
  </div>
  <div
    className="fd-mega-menu__group"
  >
    <h3
      className="fd-mega-menu__title"
    >
      Log Labels
    </h3>
    <ul
      className="fd-mega-menu__list"
    >
      <li
        className="fd-mega-menu__item"
        key="name"
      >
        <span
          aria-controls="name"
          aria-haspopup="true"
          className="fd-mega-menu__link has-child cursor-pointer"
          onClick={[Function]}
        >
          name
        </span>
        <ul
          aria-hidden={false}
          className="fd-mega-menu__sublist"
          id="name"
        >
          <li
            className="fd-mega-menu__subitem"
            key="c,d"
          >
            <span
              className="fd-mega-menu__sublink cursor-pointer"
              onClick={[Function]}
            >
              c
              d
            </span>
          </li>
        </ul>
      </li>
      <li
        className="fd-mega-menu__item"
        key="name 2"
      >
        <span
          aria-controls="name 2"
          aria-haspopup="true"
          className="fd-mega-menu__link has-child cursor-pointer"
          onClick={[Function]}
        >
          name 2
        </span>
        <ul
          aria-hidden={true}
          className="fd-mega-menu__sublist"
          id="name 2"
        >
          <li
            className="fd-mega-menu__subitem"
            key="e,f"
          >
            <span
              className="fd-mega-menu__sublink cursor-pointer"
              onClick={[Function]}
            >
              e
              f
            </span>
          </li>
        </ul>
      </li>
    </ul>
  </div>
</nav>
`;
